File: 7.Product_of_Array_Except_Self/sol.go
package main

import (
	"fmt"
	"os"
	"reflect"
	"runtime"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Product of Array Except Self (Go Version)
// ======================================================================
// Description:
// Given an integer array `nums`, return an array `answer` such that
// `answer[i]` is equal to the product of all the elements of `nums`
// except `nums[i]`.
//
// 📋 Rules:
// 1. You must write an algorithm that runs in O(n) time.
// 2. You CANNOT use the division operation.
// 3. Follow up: Can you solve it with O(1) extra space complexity?
//    (The output array does not count as extra space).
//
// 💡 Examples:
// - ProductExceptSelf([1,2,3,4])       => [24,12,8,6]
// - ProductExceptSelf([-1,1,0,-3,3])   => [0,0,9,0,0]
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)

func ProductExceptSelf1(nums []int) []int {
	n := len(nums)
	answer := make([]int, n)

	// 1. Left Pass
	answer[0] = 1
	for i := 1; i < n; i++ {
		answer[i] = answer[i-1] * nums[i-1]
	}

	// 2. Right Pass
	rightProduct := 1
	for i := n - 1; i >= 0; i-- {
		// 關鍵：將右邊累積的積「乘」進結果中
		answer[i] *= rightProduct

		// 更新右邊累積值
		rightProduct *= nums[i]
	}

	return answer
}

// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func ProductExceptSelf(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	for i := range n {
		ans[i] = 1
	}

	leftProduct := 1
	for i := range n {
		ans[i] = leftProduct * ans[i]
		leftProduct = leftProduct * nums[i]
	}

	rightProduct := 1
	for i := n - 1; i >= 0; i-- {
		ans[i] = rightProduct * ans[i]
		rightProduct = rightProduct * nums[i]
	}
	return ans
}

// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

func main() {
	runTests()
}

type TestCase struct {
	nums     []int
	expected []int
}

func runTests() {
	testCases := []TestCase{
		{[]int{1, 2, 3, 4}, []int{24, 12, 8, 6}},
		{[]int{-1, 1, 0, -3, 3}, []int{0, 0, 9, 0, 0}},
		{[]int{0, 0}, []int{0, 0}},
		{[]int{1, 0}, []int{0, 1}},
	}

	fmt.Printf("\n🧪 Testing your [ProductExceptSelf] function...\n\n")

	header := fmt.Sprintf("%-20s | %-15s | %-15s | Status", "Input nums", "Expected", "Actual")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		result := ProductExceptSelf(tc.nums)

		isMatch := reflect.DeepEqual(result, tc.expected)
		statusIcon := "✅ PASS"
		if !isMatch {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		// Helper to format slices
		fmtSlice := func(s []int) string {
			return fmt.Sprintf("%v", s)
		}

		numsStr := fmtSlice(tc.nums)
		if len(numsStr) > 18 {
			numsStr = numsStr[:15] + "..."
		}
		expStr := fmtSlice(tc.expected)
		if len(expStr) > 13 {
			expStr = expStr[:10] + "..."
		}
		resStr := fmtSlice(result)
		if len(resStr) > 13 {
			resStr = resStr[:10] + "..."
		}

		fmt.Printf("%-20s | %-15s | %-15s | %s\n",
			numsStr, expStr, resStr, statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func ProductExceptSelf(nums []int) []int {",
		"\t// TODO: Implement your solution here.",
		"\treturn []int{}",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion

File: 7.Product_of_Array_Except_Self/sol.py
import os
import sys
from typing import List

# ======================================================================
# 🧠 CHALLENGE: Product of Array Except Self (Python Version)
# ======================================================================
# Description:
# Given an integer array `nums`, return an array `answer` such that
# `answer[i]` is equal to the product of all the elements of `nums`
# except `nums[i]`.
#
# 📋 Rules:
# 1. You must write an algorithm that runs in O(n) time.
# 2. You CANNOT use the division operation.
# 3. Follow up: Can you solve it with O(1) extra space complexity?
#    (The output array does not count as extra space).
#
# 💡 Examples:
# - product_except_self([1,2,3,4])       => [24,12,8,6]
# - product_except_self([-1,1,0,-3,3])   => [0,0,9,0,0]
# ======================================================================

# region [📚 Reference Solutions] (Solutions hidden as requested)


def product_except_self1(nums: List[int]) -> List[int]:
    n = len(nums)
    answer = [1] * n

    # 1. Left Pass (你是對的)
    for i in range(1, n):
        answer[i] = answer[i - 1] * nums[i - 1]

    # 2. Right Pass (修正部分)
    right_product = 1

    # 從最後一個元素開始 (n-1)，一路走到 0
    for i in range(n - 1, -1, -1):
        # 關鍵：是「乘上去」(*=)，不是「覆蓋」(=)
        # answer[i] 目前是左邊積，right_product 是右邊積
        answer[i] *= right_product

        # 更新累積的右邊積
        right_product *= nums[i]

    return answer


# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def product_except_self(nums):
    # TODO: Implement your solution here.
    return []
# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def product_except_self(nums):\n",
        "    # TODO: Implement your solution here.\n",
        "    return []\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)
        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ([1, 2, 3, 4], [24, 12, 8, 6]),
        ([-1, 1, 0, -3, 3], [0, 0, 9, 0, 0]),
        ([0, 0], [0, 0]),  # Double zero case
        ([1, 0], [0, 1]),  # Single zero case
    ]

    print(f"\n🧪 Testing your [product_except_self] function...\n")

    header = f"{'Input nums':<25} | {'Expected':<20} | {'Actual':<20} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for nums, expected in test_cases:
        try:
            result = product_except_self(nums)
        except Exception as e:
            result = "Error"

        is_match = result == expected
        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        # Format output
        nums_str = str(nums)
        if len(nums_str) > 23:
            nums_str = nums_str[:20] + "..."
        exp_str = str(expected)
        if len(exp_str) > 18:
            exp_str = exp_str[:15] + "..."
        res_str = str(result)
        if len(res_str) > 18:
            res_str = res_str[:15] + "..."

        print(f"{nums_str:<25} | {exp_str:<20} | {res_str:<20} | {status_icon}")

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")

# endregion
